File: clients/android/app/src/main/kotlin/com/lextures/android/features/contenttools/sandbox/SandboxWebViewPool.kt
package com.lextures.android.features.contenttools.sandbox

import com.lextures.android.core.lms.ContentToolSandboxLogic

/** Caps live sandbox WebViews per screen (NFR: at most 3). Pure bookkeeping; UI owns teardown. */
object SandboxWebViewPool {
    private val alive = mutableListOf<String>()
    private var maxAlive: Int = ContentToolSandboxLogic.MAX_LIVE_WEBVIEWS

    fun configure(maxAlive: Int) {
        this.maxAlive = maxAlive
    }

    fun aliveCount(): Int = alive.size

    fun aliveIds(): List<String> = alive.toList()

    /** Registers a mount. Returns instance ids that must be torn down to stay within budget. */
    fun retain(instanceId: String): List<String> {
        alive.removeAll { it == instanceId }
        alive.add(instanceId)
        val evicted = mutableListOf<String>()
        while (ContentToolSandboxLogic.poolShouldEvict(alive.size, maxAlive)) {
            val victim = alive.removeAt(0)
            if (victim != instanceId) evicted.add(victim)
        }
        return evicted
    }

    fun release(instanceId: String) {
        alive.removeAll { it == instanceId }
    }

    fun reset() {
        alive.clear()
    }
}

File: clients/ios/Lextures/Features/ContentTools/Sandbox/SandboxWebViewPool.swift
import Foundation

/// Caps live sandbox WebViews per screen (NFR: at most 3). Pure bookkeeping; UI owns teardown.
@MainActor
final class SandboxWebViewPool {
    static let shared = SandboxWebViewPool()

    private var alive: [String] = []
    private let maxAlive: Int

    init(maxAlive: Int = ContentToolSandboxLogic.maxLiveWebViews) {
        self.maxAlive = maxAlive
    }

    var aliveCount: Int { alive.count }

    var aliveIds: [String] { alive }

    /// Registers a mount. Returns instance ids that must be torn down to stay within budget.
    @discardableResult
    func retain(_ instanceId: String) -> [String] {
        alive.removeAll { $0 == instanceId }
        alive.append(instanceId)
        var evicted: [String] = []
        while ContentToolSandboxLogic.poolShouldEvict(aliveCount: alive.count, maxAlive: maxAlive) {
            let victim = alive.removeFirst()
            if victim != instanceId {
                evicted.append(victim)
            }
        }
        return evicted
    }

    func release(_ instanceId: String) {
        alive.removeAll { $0 == instanceId }
    }

    func reset() {
        alive.removeAll()
    }
}
